src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -303,6 +303,24 @@ CreateDistributedInsertSelectPlan(Query *originalQuery,
 	return distributedPlan;
 }
 
+/*
+ * CreateInsertSelectIntoLocalTablePlan
+ */
+DistributedPlan *
+CreateInsertSelectIntoLocalTablePlan(uint64 planId, Query *originalQuery, PlannerRestrictionContext *plannerRestrictionContext)
+{
+	Query * selectQuery = ((RangeTblEntry *)lsecond(originalQuery->rtable))->subquery;
+
+	MultiTreeRoot *logicalPlan = MultiLogicalPlanCreate(selectQuery, selectQuery, plannerRestrictionContext);
+	MultiLogicalPlanOptimize(logicalPlan);
+	DistributedPlan * distPlan = CreatePhysicalDistributedPlan(logicalPlan, plannerRestrictionContext);
+
+	((RangeTblEntry *)lsecond(originalQuery->rtable))->subquery = distPlan->masterQuery;
+	distPlan->masterQuery = originalQuery;
```

Comment:
Using the master query to re-plan the insert..select query after replacing the select with the master query is a very interesting idea.

Unfortunately, router plannable queries do not have a master query (will be NULL). However, I think it might be possible to construct a synthetic master query that simply selects all the results for router queries. You'd need to follow (some of) the steps in `BuildJobQuery` and specifically `QueryJoinTree()` under `case T_MultiCollect`.